### ISSUE TITLE
Improve support for is_file and family

### ISSUE THREAD

**Issue by villfa** [vimeo/psalm#7770]
Hi,

The functions that expect a valid path (`file_exists`, `is_readable`, ...) always return *false* when an empty string is provided ( https://3v4l.org/TcecN ).
So when those functions return *true* we can say that the argument is a non-empty string.

If my assumption is right it could help with situations like this:
https://psalm.dev/r/26f209fe21

**Comment by AndrolGenhald** [comment]
Sounds nice! Care for a PR? You should be able to do this with [`@psalm-assert-if-true`](https://psalm.dev/docs/annotating_code/adding_assertions/) in CoreGenericFunctions.phpstub.

**Comment by villfa** [comment]
Thanks for the pointer!
I don't promise anything but I'll take a look when I have time.

**Comment by villfa** [comment]
I've started to work on this and I'm wondering what should I do with the functions like `mkdir` because they emit a warning when the argument is an empty string.
Example:
```php
<?php var_dump(mkdir(''));
// Warning: mkdir(): No such file or directory in test.php on line 1
// bool(false)
```

I think that modifying the signature by requiring a non-empty string would be a better approach. WDYT?

**Comment by AndrolGenhald** [comment]
That sounds good to me, if it emits a warning when given an empty string make it take a non-empty-string, if it just returns false with no warning use an assertion.

**Comment by villfa** [comment]
Here some oddities I've found while working on it:
- https://psalm.dev/r/af13a552ce
- https://psalm.dev/r/d2819c2e06

**Comment by orklah** [comment]
First one is indeed a bug. `string` - `non-empty-string` should still be some kind of string, not `never`.

The second one is legit. Your psalm-retun-if-false is redundant

**Comment by AndrolGenhald** [comment]
@orklah Yeah the second one is redundant, but check out the error message: `Type string for $s is always !string`

**Comment by orklah** [comment]
oh right, the message should probably not be negated here

**Comment by AndrolGenhald** [comment]
I think it's more than just being incorrectly negated though, [this](https://psalm.dev/r/d7795c3a47) doesn't make sense at all.

**Comment by orklah** [comment]
yeah, that's a weird one
